### ISSUE TITLE
Update required fields according to document

### ISSUE THREAD

**Issue by jamstooks** [AASHE/hub#140]
All submission forms should be updated according to the [Hub Fields by Content Type](https://docs.google.com/spreadsheets/d/1ADsnCZXKU1lhA4GHc2xhsY5ddLRi7kgvYJLSTHZKlmw/edit#gid=935008516) document.

- [x] Academic Programs
- [x] Case Studies
- [x] Centers & Institutes
- [x] Conference Pres.
- [x] Publications
- [x] Photos
- [x] Course Materials
- [x] Tools
- [x] Videos
- [x] Outreach Materials

This should consolidate a number of tickets.

**Comment by jamstooks** [comment]
@bartTC,

Each content type has different requirements for metadata. For example:
 - a photo only requires an Image, but no Author or Files.
 - a conference presentation requires a file upload, but limits to 3
 - a publication requires either a document or a website

Given the new structure of the forms, I think it won't be too tricky to update the form validation. I'm leaning towards this solution to handle all the variations on requirements for metadata.

Adding a `metadata_requirements` classmethod to the ContentTypes.
 
Each method will return a simple dictionary. Here's an example for Photographs:

    {
        images: {max: 5, min: 1},  # one required, limit 5
        files: {max: 0, min: 0},  # not included
        authors: {max: 0, min: 0},  # authors are not included
        websites: {max: 5, min: 0},  # optional, up to 5
    }
    
and Course Materials with the addition of `conditionally_required`:

    {
        images: {max: 0, min: 0},  # not included
        files: {max: 5, min: 0},  # optional, up to 5
        authors: {max: 5, min: 0},  # optional, up to 5
        websites: {max: 5, min: 0},  # optional, up to 5,
        # finally, at least one file or website is required
        conditionally_required: {'files', 'websites'}
    }

In fact, I could even leave out the keys that don't apply for a specific content type.

`SubmitFormView.get_form` would simply refer to `metadata_requirements` to determine values for min/max/extra for formsets, and only include the relevant formsets. It would also update the validation to handle `conditionally_required` where necessary.

This also has the added benefit of reducing duplication, but pulling the image fields out of the primary form at the top and pulling them back into the formset.

I'm going to start on this, but thought it would be helpful to document it first and see if you had any thoughts...

**Comment by bartTC** [comment]
Thought here:

1) For optional parts, like `files max:0` you don't do any validation. You simply don't render it in the template. They are optional anyway. So all other restrictions apply too.

2) For conditional validation. Don't do something like `conditionally_required: {'files', 'websites'}` and some magic afterwards. It's not flexible enough.

Rather create a custom form class for each content type, which inherits form the regular Form we use now, and use that if exists. Similar how we handle AcademicProgram Filtersets right now. 

https://github.com/AASHE/hub/blob/master/hub/apps/content/types/academic.py#L57-L60
https://github.com/AASHE/hub/blob/master/hub/apps/browse/views.py#L117-L129

**Comment by jamstooks** [comment]
I took a stab at this in a new branch, [required-optional-fields-issue-140](https://github.com/AASHE/hub/tree/required-optional-fields-issue-140). It's tested and working, but I'd definitely welcome feedback.

The reason I did the validation in the view was that the inline forms weren't available within the `document_form`... they were only in the view.

I also stuck with `conditionally_required` in the requirements dictionary for now. If we need more advanced validation logic down the road for specific resources, I can break out the forms.
